src/gateway.c

```diff
@@ -58,8 +58,8 @@ void gateway__leader_close(struct gateway *g, int reason)
 			struct raft_barrier *b = &g->leader->exec->barrier.req;
```

Comment:
This branch is no longer covered by the tests, can you check why it's the case? Maybe it's no longer needed. I'm being nitty about it because there were issues with the cleanup in the past.

Reply:
This change seems indeed a bit fishy. I'd like to understand why it's done.

I'm realizing that putting stuff in the `req` object is a bit more subtle than I thought. We probably need a bit more clarity around that, but I didn't think about it too much yet.

Reply:
@MathieuBordere So the gateway uses `leader__exec` in two places: `handle_exec` and `handle_exec_sql`. With this change, the `leader__exec` call from `handle_exec_sql` never installs a barrier because we have a new call to `leader__barrier` just before `sqlite3_prepare_v2`. It sounds like we don't have any test coverage of `handle_exec` that results in actually installing a barrier, but I can quickly fix that.

Reply:
Okay, would be nice to have that!

Reply:
@freeekanayaka Just to clarify "this change", you mean the commit updating `gateway__leader_close`?

Reply:
Exactly.

Reply:
Added that unit test in the latest commit -- I've confirmed with `make check-code-coverage` that our tests now cover all the branches in `gateway__leader_close`.

@freeekanayaka My reasoning is that with the other changes in this PR, it's now possible to have an in-flight barrier even when `g->stmt == NULL` (for example, if `handle_prepare` was called). The check in `gateway__leader_close` needs to be updated so that the barrier will be properly cancelled in that situation. The `g->req != NULL` check determines whether there's a request in flight, and then if there is we check which fields (regular barrier, exec barrier, prepared statement) need to be cleaned up. I think the only way this change could cause problems is if it's possible to have `g->req == NULL` but `g->stmt != NULL` -- I haven't found a way for that to happen, but it's possible I missed something.

Reply:
Ok, I'm not sure either, but since you've looked at it and you added test I believe it's reasonably safe.

This part of the code is unfortunate because the real solution would be to make `gateway__close()` asynchronous, since that's what it is. We've talked about it in the past, but it's a change that requires a bit of rework at various layers, so for now it's still pending.